### PR TITLE
Feature: `@knownValues` decorator 

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix https://github.com/Azure/cadl-azure/issues/1151

Adds `@knownValues` decorator
```cadl
 enum KnownPetType {
    Dog, Cat
  }

  @knownValues(KnownPetType)
  model PetType extends string {}
```

Which produce the following OpenAPI:
```json5
{
  oneOf: [
    { type: "string" }, 
    { type: "string", enum: ["Dog", "Cat"] }
  ],
}
```

Additional changes:
- Added helper to validate decorator target is intrinsic type

### CODE REVIEW

**Comment by bterlson** [comment]
Should this work for `model is string`, and both extends/is with `number`?

**Comment by timotheeguerin** [comment]
@bterlson I was under the impression `model is string` shouldn't be allowed as this issue specify  #235. Let me check for numbers

**Comment by bterlson** [comment]
Ok. I don't think I was present for that decision, anyone remember the rationale? It seems like something that has a natural representation in swagger, and it seems convenient to have a type that is a string or number with known values

File: packages/compiler/lib/decorators.ts [hunk 1]
```diff
@@ -546,3 +516,32 @@ export function $friendlyName(
...
+    !validateDecoratorTargetIntrinsic(context.program, target, "@knownValues", [
+      "string",
+      "int8",
+      "int16",
+      "int32",
+      "int64",
+      "float32",
+      "float64",
+    ]) ||
+    !validateDecoratorParamType(context.program, target, knownValues, "Enum")
```
Comment (markcowl) [inline packages/compiler/lib/decorators.ts:537]: Can we do a deeper evaluation here, to ensure the base type of the enum options match the intrinsic type of the target?

**Review by markcowl — CHANGES_REQUESTED** [review]
Also, I think as part of this, we want to change the model type properties that use enums  in the Azure samples to use KnownValues with those enums instead.

File: packages/compiler/lib/decorators.ts (hunk 1 above)
Comment (timotheeguerin) [reply packages/compiler/lib/decorators.ts:539]: added that validation, this becomes closer and closer to needing having a full type assignment checker with subtype/supertype relations

**Review by markcowl — APPROVED** [review]
The other part of this will be updating the cadl-azure repo for similar behavior - using x-ms-enum only for KnownValues and using a strct enum for conventional enum schemas.  Note that all Azure specs (here, ProviderHub specs) should use *exclusively* enums.

**Pull request merged**